renku/core/models/provenance/datasets.py

```diff
@@ -0,0 +1,515 @@
+        """Remove a dataset."""
+        revision = revision or "HEAD"
+        date = date or datetime.datetime.now(datetime.timezone.utc)
+        current_dataset = self.get(dataset.identifier)
+        if not current_dataset:
+            communication.warn(f"Cannot find dataset to delete `{dataset.identifier}` at revision `{revision}`")
+            return
+        assert not current_dataset.is_deleted(), f"Dataset `{current_dataset.name}` was deleted before."
+        current_dataset.date_deleted = date
+
+    @classmethod
+    def from_json(cls, path):
+        """Return an instance from a file."""
+        if Path(path).exists():
+            with open(path) as file_:
+                data = json.load(file_)
+                self = cls.from_jsonld(data=data) if data else DatasetProvenance()
+        else:
+            self = DatasetProvenance()
+
+        self._path = path
+
+        return self
+
+    @classmethod
+    def from_jsonld(cls, data):
+        """Create an instance from JSON-LD data."""
+        if isinstance(data, cls):
+            return data
+        elif not isinstance(data, list):
+            raise ValueError(data)
+
+        return DatasetProvenanceSchema(flattened=True).load(data)
+
+    def to_json(self, path=None):
+        """Write to file."""
+        path = path or self._path
+        data = self.to_jsonld()
+        with open(path, "w", encoding="utf-8") as file_:
+            json.dump(data, file_, ensure_ascii=False, sort_keys=True, indent=2)
+
+    def to_jsonld(self):
+        """Create JSON-LD."""
+        return DatasetProvenanceSchema(flattened=True).dump(self)
+
+
+class NewDatasetFileSchema(JsonLDSchema):
+    """DatasetFile schema."""
+
+    class Meta:
+        """Meta class."""
+
+        rdf_type = [prov.Entity, schema.DigitalDocument]
+        model = DatasetFile
+        unknown = EXCLUDE
+
+    based_on = Nested(schema.isBasedOn, DatasetFileSchema, missing=None, propagate_client=False)
+    date_added = DateTimeList(schema.dateCreated, format="iso", extra_formats=("%Y-%m-%d",))
+    date_deleted = fields.DateTime(prov.invalidatedAtTime, missing=None, allow_none=True, format="iso")
+    entity = Nested(renku.entity, EntitySchema, missing=None)
+    id = fields.Id(init_name="id")
+    is_external = fields.Boolean(renku.external, missing=False)
+    source = fields.String(renku.source, missing=None)
+    url = fields.String(schema.url, missing=None)
+
+
+class NewDatasetSchema(JsonLDSchema):
+    """Dataset schema."""
+
+    class Meta:
+        """Meta class."""
+
+        rdf_type = schema.Dataset
+        model = Dataset
+        unknown = EXCLUDE
+
+    creators = Nested(schema.creator, PersonSchema, many=True)
+    date_created = fields.DateTime(
+        schema.dateCreated, missing=None, allow_none=True, format="iso", extra_formats=("%Y-%m-%d",),
+    )
+    date_deleted = fields.DateTime(prov.invalidatedAtTime, missing=None, allow_none=True, format="iso")
```

Comment:
If we want to be consistent with PROV-O, then to use `invalidatedAtTime`, dataset also needs to be a `prov.entity`, as that's the domain of that property.